library/psa_crypto.c

```diff
@@ -3395,6 +3395,440 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
+                              size_t *ciphertext_length,
+                              uint8_t *tag,
+                              size_t tag_size,
+                              size_t *tag_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    *ciphertext_length = 0;
+    *tag_length = tag_size;
+
+    status = psa_aead_final_checks( operation );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    if( !operation->is_encrypt )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_aead_finish( operation, ciphertext,
+                                             ciphertext_size,
+                                             ciphertext_length,
+                                             tag, tag_size, tag_length );
+
+exit:
+    /* In case the operation fails and the user fails to check for failure or
+     * the zero tag size, make sure the tag is set to something impossible.
+     * Even if the operation succeeds, make sure we set the rest of the
+     * buffer to something impossible to prevent potential leakage of
+     * anything previously placed in the same buffer.*/
+    if( status != PSA_SUCCESS )
+        memset( tag, '!', tag_size );
+    else if( *tag_length < tag_size )
+        memset( tag + *tag_length, '!', ( tag_size - *tag_length ) );
+
+    psa_aead_abort( operation );
+
+    return( status );
+}
+
+/* Finish authenticating and decrypting a message in a multipart AEAD
+   operation.*/
+psa_status_t psa_aead_verify( psa_aead_operation_t *operation,
+                              uint8_t *plaintext,
+                              size_t plaintext_size,
+                              size_t *plaintext_length,
+                              const uint8_t *tag,
+                              size_t tag_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    *plaintext_length = 0;
+
+    status = psa_aead_final_checks( operation );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    if( operation->is_encrypt )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_aead_verify( operation, plaintext,
+                                             plaintext_size,
+                                             plaintext_length,
+                                             tag, tag_length );
+
+exit:
+    psa_aead_abort( operation );
+
+    return( status );
+}
+
+/* Abort an AEAD operation. */
+psa_status_t psa_aead_abort( psa_aead_operation_t *operation )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if( operation->id == 0 )
+    {
+        /* The object has (apparently) been initialized but it is not (yet)
+         * in use. It's ok to call abort on such an object, and there's
+         * nothing to do. */
+        return( PSA_SUCCESS );
+    }
+
+    status = psa_driver_wrapper_aead_abort( operation );
+
+    operation->id = 0;
```

Comment:
Why not just
```C
mbedtls_platform_zeroize( operation, sizeof( *operation ) );
```
?

Reply:
Should we be using platform_zeroize here (its slow) rather than just memset? There are are no secrets per se in the operation.

Reply:
indeed, ok for memset.

Reply:
The operation can contain sensitive data! Typically an `mbedtls_cipher_context`, which contains the secret key.

Reply:
There is a call to `psa_driver_wrapper_aead_abort` that can/could take care of such sensitive data but we probably better go for belt and braces here and use `mbedtls_platform_zeroize()` here. Should we revisit this in the other `xyz_abort()` as well (not in this PR though) ?

Reply:
Actually I realize I made a mistake in my previous comment: the operation contains sensitive data, but at this point `mbedtls_cipher_free` should have already been called, so all that remains to be wiped is metadata. This doesn't need to be zeroized, just reset in case the operation object is reused, and for that memset is fine.

This leaves the question open: are we comfortable assuming that drivers will zeroize memory? Or do we want to zeroize in the core to be sure?

Either way, you're right, we should be consistent for all types of algorithms.

Reply:
Left as memset for now, if we want zeroize, then we should probably do all algs at the same time.

Reply:
Filed as https://github.com/ARMmbed/mbedtls/issues/4956